### PR TITLE
Add system-characteristics 'has-authorization-boundary' constraints

### PR DESCRIPTION
**Pull request by Gabeblis**
# Committer Notes

### Description 
Added system-characteristics 'has-authorization-boundary' constraints and tests for each of the constraints. These should be added as they are part of the effort write all of the constraints from the constraint tracker.

### Changes  
- Added constraints has-authorization-boundary-diagram, has-authorization-boundary-diagram-uuid, has-authorization-boundary-diagram-description, has-authorization-boundary-diagram-link, has-authorization-boundary-diagram-caption, has-authorization-boundary-diagram-link-rel, and has-authorization-boundary-diagram-link-rel-allowed-value to fedramp-external-constraints.xml.  
- Added pass and fail yaml tests for all of the above constraints.  
- Edited ssp-all-VALID.xml to ensure all constraints pass correctly.  

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
~~- [ ] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?~~
~~- [ ] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?~~

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 1]
```diff
@@ -62,6 +62,27 @@
             <expect id="categorization-has-information-type-id" target="//system-characteristics" test="system-information/information-type/categorization/information-type-id" level="ERROR">
             <message>A FedRAMP SSP information type categorization must have at least one information type identifier.</message>
             </expect>
+            <expect id="has-authorization-boundary-diagram" target="./system-characteristics/authorization-boundary" test="diagram" level="WARNING">
+            <message>A FedRAMP SSP has at least one authorization boundary diagram.</message>
+            </expect>
+            <expect id="has-authorization-boundary-diagram-uuid" target="./system-characteristics/authorization-boundary" test="diagram/@uuid" level="ERROR">
+            <message>Each OSCAL SSP authorization boundary diagram has a unique identifier.</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:70]: Blocking: is there a unique requirement per UUID? If so, we want to check that each UUID is unique with an index, not just that it exists. This feedback may require a little more complex of a suggestion than inline in this area of the diff, so I will find some other way to workshop it later.

Also, we should format and indent in either way.

```suggestion
            <expect id="has-authorization-boundary-diagram-uuid" target="./system-characteristics/authorization-boundary" test="diagram/@uuid" level="ERROR">
                <message>Each authorization boundary diagram in FedRAMP SSP must have a unique identifier.</message>
            </expect>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 2]
```diff
@@ -62,6 +62,27 @@
...
+            <expect id="has-authorization-boundary-diagram-description" target="./system-characteristics/authorization-boundary" test="diagram/description" level="ERROR">
+            <message>An OSCAL SSP document authorization boundary diagram has a description.</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:73]:
```suggestion
            <expect id="has-authorization-boundary-diagram-description" target="./system-characteristics/authorization-boundary" test="diagram/description" level="ERROR">
                <message>An OSCAL SSP document authorization boundary diagram has a description.</message>
            </expect>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 3]
```diff
@@ -62,6 +62,27 @@
...
+            <expect id="has-authorization-boundary-diagram-link" target="./system-characteristics/authorization-boundary" test="diagram/link" level="ERROR">
+            <message>Each FedRAMP SSP authorization boundary diagram has a link.</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:76]:
```suggestion
            <expect id="has-authorization-boundary-diagram-link" target="./system-characteristics/authorization-boundary" test="diagram/link" level="ERROR">
                <message>Each FedRAMP SSP authorization boundary diagram has a link.</message>
            </expect>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 4]
```diff
@@ -62,6 +62,27 @@
...
+            <expect id="has-authorization-boundary-diagram-link-rel" target="./system-characteristics/authorization-boundary" test="diagram/link/@rel" level="ERROR">
+            <message>Each FedRAMP SSP authorization boundary diagram has a link rel attribute.</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:79]:
```suggestion
            <expect id="has-authorization-boundary-diagram-link-rel" target="./system-characteristics/authorization-boundary" test="diagram/link/@rel" level="ERROR">
                <message>Each FedRAMP SSP authorization boundary diagram has a link rel attribute.</message>
            </expect>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 5]
```diff
@@ -62,6 +62,27 @@
...
+            <expect id="has-authorization-boundary-diagram-link-rel-allowed-value" target="./system-characteristics/authorization-boundary" test="diagram/link/@rel eq 'diagram'" level="ERROR">
+            <message>Each FedRAMP SSP authorization boundary diagram has a link rel attribute with the value "diagram".</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:82]:
```suggestion
            <expect id="has-authorization-boundary-diagram-link-rel-allowed-value" target="./system-characteristics/authorization-boundary" test="diagram/link/@rel eq 'diagram'" level="ERROR">
                <message>Each FedRAMP SSP authorization boundary diagram has a link rel attribute with the value "diagram".</message>
            </expect>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 6]
```diff
@@ -62,6 +62,27 @@
...
+            <expect id="has-authorization-boundary-diagram-caption" target="./system-characteristics/authorization-boundary" test="diagram/caption" level="ERROR">
+            <message>Each FedRAMP SSP authorization boundary diagram has a caption.</message>
+            </expect>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:85]:
```suggestion
            <expect id="has-authorization-boundary-diagram-caption" target="./system-characteristics/authorization-boundary" test="diagram/caption" level="ERROR">
                <message>Each FedRAMP SSP authorization boundary diagram has a caption.</message>
            </expect>
```

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]
Excellent work, just two significant (but not super bad) issues: one issue I want to run by internal review staff, the other I think we need an index but can't just insert that with a diff and need to actually use this cool new `mq` thing to figure it out.

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 1 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml:70]: > Blocking: is there a unique requirement per UUID? If so, we want to check that each UUID is unique with an index, not just that it exists. This feedback may require a little more complex of a suggestion than inline in this area of the diff, so I will find some other way to workshop it later.
> 
> Also, we should format and indent in either way.

Thank you for catching the formatting. I believe this specific constraint is simply checking that a diagram has a '@uuid' attribute, not necessarily validating if the uuid itself is unique within the SSP. I'll have to take a look and see if there is some other constraint that checks if each uuid is unique.

Comment (aj-stein-gsa) [reply src/validations/constraints/fedramp-external-constraints.xml:70]: Ok I'll review the spreadsheet and ask about the boundary diagram issue still, expect an approval with formatting tonight or tomorrow morning then. 😁

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 7]
```diff
@@ -62,6 +62,27 @@
             <expect id="categorization-has-information-type-id" target="//system-characteristics" test="system-information/information-type/categorization/information-type-id" level="ERROR">
             <message>A FedRAMP SSP information type categorization must have at least one information type identifier.</message>
             </expect>
+            <expect id="has-authorization-boundary-diagram" target="./system-characteristics/authorization-boundary" test="diagram" level="WARNING">
+            <message>A FedRAMP SSP has at least one authorization boundary diagram.</message>
+            </expect>
```
[the remark this answers is not in the document]
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml]: > Formatting and wording recommendation:
> 
> Another consideration: can a SSP have more than one diagram, like a multipart diagram or something? I will ask the reviewer teams if that ever happens. I would think that changes the `@level` and the `@test` if this Metapath works or we need a `test="count(diagram) >= 1` alternative.

There can definitely be multiple diagrams within `system-characteristics`. For example, there could be one within `authorization-boundary`, another within `network-architecture`,  another within `data-flow` and so on. This target specifically targets the `authorization-boundary` and is sufficient for checking the existence of at least one `diagram`. There are other constraints that that check the existence of `diagram` within those other elements that I mentioned above. I don't think a change is needed here.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 8]
```diff
@@ -57,19 +57,40 @@
...
-            <message>This FedRAMP SSP does define its NIST SP 800-63 identity assurance level (IAL).</message>
+                <message>This FedRAMP SSP does define its NIST SP 800-63 identity assurance level (IAL).</message>
             </expect>
             <expect id="has-authenticator-assurance-level" target="./system-characteristics" test="prop[@name eq 'authenticator-assurance-level']" level="INFORMATIONAL">
-            <message>This FedRAMP SSP does define its NIST SP 800-63 authenticator assurance level (AAL).</message>
+                <message>This FedRAMP SSP does define its NIST SP 800-63 authenticator assurance level (AAL).</message>
             </expect>
             <expect id="has-federation-assurance-level" target="./system-characteristics" test="prop[@name eq 'federation-assurance-level']" level="INFORMATIONAL">
-            <message>This FedRAMP SSP does define its NIST SP 800-63 federation assurance level (IAL).</message>
+                <message>This FedRAMP SSP does define its NIST SP 800-63 federation assurance level (IAL).</message>
```
Comment (aj-stein-gsa) [inline src/validations/constraints/fedramp-external-constraints.xml:72]: I missed this errant acronym in my copy-paste feedback yesterday and will merge it and approve.

```suggestion
                <message>This FedRAMP SSP does define its NIST SP 800-63 federation assurance level (FAL).</message>
```

**Review by aj-stein-gsa — DISMISSED** [review]

**Review by aj-stein-gsa — APPROVED** [review]

**Pull request merged**